src/nnet3/nnet-simple-component.cc

```diff
@@ -4261,6 +4254,256 @@ std::string MaxpoolingComponent::Info() const {
+  KALDI_ASSERT((input_z_dim_ - pool_z_size_) % pool_z_step_  == 0);
+}
+
+// initialize the component using configuration file
+void Maxpooling3dComponent::InitFromConfig(ConfigLine *cfl) {
+  int32 input_x_dim = -1, input_y_dim = -1, input_z_dim = -1;
+  int32 pool_x_size = -1, pool_y_size = -1, pool_z_size = -1;
+  int32 pool_x_step = -1, pool_y_step = -1, pool_z_step = -1;
+  bool ok = true;
+
+  ok = ok && cfl->GetValue("input-x-dim", &input_x_dim);
+  ok = ok && cfl->GetValue("input-y-dim", &input_y_dim);
+  ok = ok && cfl->GetValue("input-z-dim", &input_z_dim);
+  ok = ok && cfl->GetValue("pool-x-size", &pool_x_size);
+  ok = ok && cfl->GetValue("pool-y-size", &pool_y_size);
+  ok = ok && cfl->GetValue("pool-z-size", &pool_z_size);
+  ok = ok && cfl->GetValue("pool-x-step", &pool_x_step);
+  ok = ok && cfl->GetValue("pool-y-step", &pool_y_step);
+  ok = ok && cfl->GetValue("pool-z-step", &pool_z_step);
+
+  if (cfl->HasUnusedValues())
+    KALDI_ERR << "Could not process these elements in initializer: "
+              << cfl->UnusedValues();
+  if (!ok)
+    KALDI_ERR << "Bad initializer " << cfl->WholeLine();
+  Init(input_x_dim, input_y_dim, input_z_dim,
+       pool_x_size, pool_y_size, pool_z_size,
+       pool_x_step, pool_y_step, pool_z_step);
+}
+
+/*
+  3d max pooling propagate function
+  It is assumed that each row of the input matrix
+  is a vectorized 3D-tensor in zxy format.
+  This function first performs max pooling along the x and y axes
+  due to the convinience of the input vectorization format.
+  It then performs max pooling along the z axis if necessary.
+  The output matrix is also in zxy vectorization format.
+*/
+
+void Maxpooling3dComponent::Propagate(const ComponentPrecomputedIndexes *indexes,
+                                    const CuMatrixBase<BaseFloat> &in,
+                                    CuMatrixBase<BaseFloat> *out) const {
+  int32 num_frames = in.NumRows();
+  int32 num_pools_x = 1 + (input_x_dim_ - pool_x_size_) / pool_x_step_;
+  int32 num_pools_y = 1 + (input_y_dim_ - pool_y_size_) / pool_y_step_;
+  int32 num_pools_z = 1 + (input_z_dim_ - pool_z_size_) / pool_z_step_;
+
+  // Do the max-pooling first along x and y axis
+  CuMatrix<BaseFloat> patches_zyx(num_frames, num_pools_x * num_pools_y * input_z_dim_, kUndefined);
+  for (int32 qx = 0; qx < num_pools_x; qx++) {
+    for (int32 qy = 0; qy < num_pools_y; qy++) {
+      // get output buffer of the pool
+      int32 q = qy + qx * num_pools_y;
+      CuSubMatrix<BaseFloat> pool(patches_zyx.ColRange(q * input_z_dim_, input_z_dim_));
+      pool.Set(-1e20); // reset a large negative value
+      int32 offset_x = qx * pool_x_step_;
+      int32 offset_y = qy * pool_y_step_;
+      for (int32 px = offset_x; px < (pool_x_size_ + offset_x); px++) {
+        for (int32 py = offset_y; py < (pool_y_size_ + offset_y); py++) {
+          int32 p = py + px * input_y_dim_;
+          pool.Max(in.ColRange(p * input_z_dim_, input_z_dim_));
```

Comment:
er... you seem to have a very tight inner loop here: the operation nested is inside 3 loops!  I have a hard time believing this could be efficient on a GPU.